File: lib/src/tokenizer.dart
import 'dart:math';

import 'extension.dart';

class Tokenizer {
  final int minChars;
  final int maxChars;
  final String delimiter;
  final bool caseSensitive;
  final bool forwardedOnly;

  const Tokenizer(
      {this.minChars = 3,
      this.maxChars = 64,
      this.delimiter = ' ',
      this.caseSensitive = false,
      this.forwardedOnly = true});

  Set<String> process(String input) {
    final words = input.split(delimiter);
    return words.flatMap(_processWord).toSet();
  }

  Iterable<String> _processWord(String word) {
    final element = caseSensitive ? word : word.toLowerCase();
    final results = <String>{};
    final effectiveMaxChars = min(maxChars, element.length);
    for (int charCount = minChars;
        charCount <= effectiveMaxChars;
        charCount++) {
      if (forwardedOnly) {
        results.add(element.substring(0, charCount));
        continue;
      }

      int forwardedPos = 0;
      int reversedPos = element.length;
      while (forwardedPos + charCount <= reversedPos) {
        if (forwardedPos + charCount <= reversedPos) {
          results
              .add(element.substring(forwardedPos, forwardedPos + charCount));
        }
        // when reversedPos - charCount == forwardPos, we already added above.
        if (reversedPos - charCount > forwardedPos) {
          results.add(element.substring(reversedPos - charCount, reversedPos));
        }
        forwardedPos++;
        reversedPos--;
      }
    }
    return results;
  }
}

File: js/src/tokenizer.ts
import _ from 'lodash';

class Tokenizer {
    minChars: number;
    maxChars: number;
    delimiter: string;
    caseSensitive: boolean;

    constructor({ minChars = 3, maxChars = 64, delimiter = ' ', caseSensitive = false }) {
        this.minChars = minChars;
        this.maxChars = maxChars;
        this.delimiter = delimiter;
        this.caseSensitive = caseSensitive;
    }

    process(input: string) {
        const words = input.split(this.delimiter);
        return _.uniq(_.flatMap(words, (word) => this._processWord(word)));
    }

    _processWord(word: string): Array<string> {
        const element = this.caseSensitive ? word : word.toLowerCase();
        const results = new Array<string>();
        const effectiveMaxChars = Math.min(this.maxChars, element.length);

        for (let charCount = this.minChars; charCount <= effectiveMaxChars; charCount++) {
            let forwardedPos = 0;
            let reversedPos = element.length;

            while (forwardedPos + charCount <= reversedPos) {
                if (forwardedPos + charCount <= reversedPos) {
                    results.push(element.substring(forwardedPos, forwardedPos + charCount));
                }

                if (reversedPos - charCount > forwardedPos) {
                    results.push(element.substring(reversedPos - charCount, reversedPos));
                }

                forwardedPos++;
                reversedPos--;
            }
        }

        return results;
    }
}

export default Tokenizer;
